File: test/deploy-config.test.mjs
import { describe, expect, it } from "vitest";
import {
  buildDeployConfig,
  buildD1DatabasesOverride,
  buildKvNamespacesOverride,
  buildR2BucketsOverride,
  buildQueuesOverride,
} from "../src/deploy-config.mjs";

const base = {
  $schema: "./node_modules/wrangler/config-schema.json",
  main: "./src/index.ts",
  compatibility_date: "2026-08-08",
  observability: { enabled: true },
  vars: { SHOULD_NOT_SURVIVE: "the committed vars are never carried forward as-is" },
  env: { production: { vars: { SECRET: "prod value" } } },
};

describe("buildDeployConfig", () => {
  it("carries forward only the safe structural keys plus the given overrides", () => {
    const result = buildDeployConfig(base, { name: "blue-honey-badger-12345", vars: { X: "1" } });
    expect(result).toEqual({
      $schema: base.$schema,
      main: base.main,
      compatibility_date: base.compatibility_date,
      observability: base.observability,
      name: "blue-honey-badger-12345",
      vars: { X: "1" },
    });
  });

  it("never carries forward the committed vars or env blocks", () => {
    const result = buildDeployConfig(base, { name: "n", vars: {} });
    expect(result.vars).toEqual({});
    expect(result.env).toBeUndefined();
  });

  it("includes hyperdrive only when explicitly passed", () => {
    const withoutHyperdrive = buildDeployConfig(base, { name: "n", vars: {} });
    expect(withoutHyperdrive.hyperdrive).toBeUndefined();

    const withHyperdrive = buildDeployConfig(base, {
      name: "n",
      vars: {},
      hyperdrive: [{ binding: "HYPERDRIVE", id: "abc123" }],
    });
    expect(withHyperdrive.hyperdrive).toEqual([{ binding: "HYPERDRIVE", id: "abc123" }]);
  });

  // A Durable Object class lives inside the deployed script itself, not a
  // separately provisioned resource — a fresh Worker name automatically gets
  // fresh DO storage. Nothing to provision, nothing shared with production.
  it("always carries durable_objects and migrations forward, no opt-in required", () => {
    const configWithDO = {
      ...base,
      durable_objects: { bindings: [{ name: "ROOM", class_name: "Room" }] },
      migrations: [{ tag: "v1", new_sqlite_classes: ["Room"] }],
    };
    const result = buildDeployConfig(configWithDO, { name: "n", vars: {} });
    expect(result.durable_objects).toEqual(configWithDO.durable_objects);
    expect(result.migrations).toEqual(configWithDO.migrations);
  });

  // This is the actual bug a security audit found: the first version of `el`
  // deployed an ephemeral Worker with live access to whatever D1/KV/R2/queue
  // bindings the committed config declared, because it only ever overwrote
  // name/vars/hyperdrive and left everything else untouched.
  it.each(["d1_databases", "kv_namespaces", "r2_buckets", "queues", "ai", "vectorize", "services"])(
    "refuses to silently deploy with a %s binding present",
    (key) => {
      const configWithBinding = { ...base, [key]: [{ binding: "SOMETHING" }] };
      expect(() => buildDeployConfig(configWithBinding, { name: "n", vars: {} })).toThrow(
        new RegExp(key),
      );
    },
  );

  it("carries a stateful binding forward only with explicit opt-in", () => {
    const configWithD1 = { ...base, d1_databases: [{ binding: "DB", database_id: "abc" }] };
    const result = buildDeployConfig(configWithD1, {
      name: "n",
      vars: {},
      unsafeInheritBindings: true,
    });
    expect(result.d1_databases).toEqual([{ binding: "DB", database_id: "abc" }]);
  });

  it("never carries forward routes or triggers, even with unsafeInheritBindings", () => {
    const configWithRoutes = {
      ...base,
      routes: [{ pattern: "example.com/*" }],
      triggers: { crons: ["0 0 * * *"] },
    };
    const result = buildDeployConfig(configWithRoutes, {
      name: "n",
      vars: {},
      unsafeInheritBindings: true,
    });
    expect(result.routes).toBeUndefined();
    expect(result.triggers).toBeUndefined();
  });

  // The point of this feature: when `el` has already provisioned a fresh,
  // environment-scoped resource, an override replaces the committed
  // (production-pointing) binding entirely — no opt-in needed, because
  // there's nothing production-pointing left in the deployed config.
  it("uses a provided override instead of throwing, with no opt-in required", () => {
    const configWithD1 = { ...base, d1_databases: [{ binding: "DB", database_id: "prod-id" }] };
    const freshD1 = [{ binding: "DB", database_name: "n-api-db", database_id: "fresh-id" }];
    const result = buildDeployConfig(configWithD1, {
      name: "n",
      vars: {},
      overrides: { d1_databases: freshD1 },
    });
    expect(result.d1_databases).toEqual(freshD1);
  });

  it("only overrides the specific stateful keys given, still gating the rest", () => {
    const configWithBoth = {
      ...base,
      d1_databases: [{ binding: "DB" }],
      kv_namespaces: [{ binding: "CACHE" }],
    };
    expect(() =>
      buildDeployConfig(configWithBoth, {
        name: "n",
        vars: {},
        overrides: { d1_databases: [{ binding: "DB", database_id: "fresh" }] },
      }),
    ).toThrow(/kv_namespaces/);
  });
});

describe("buildD1DatabasesOverride", () => {
  it("maps each declared binding to its provisioned id and name", () => {
    const result = buildD1DatabasesOverride([{ binding: "DB" }], {
      DB: { id: "uuid-1", name: "n-api-db" },
    });
    expect(result).toEqual([{ binding: "DB", database_name: "n-api-db", database_id: "uuid-1" }]);
  });

  it("handles multiple bindings independently", () => {
    const result = buildD1DatabasesOverride([{ binding: "DB" }, { binding: "ANALYTICS" }], {
      DB: { id: "uuid-1", name: "n-api-db" },
      ANALYTICS: { id: "uuid-2", name: "n-api-analytics" },
    });
    expect(result).toHaveLength(2);
    expect(result[1]).toEqual({
      binding: "ANALYTICS",
      database_name: "n-api-analytics",
      database_id: "uuid-2",
    });
  });
});

describe("buildKvNamespacesOverride", () => {
  it("maps each declared binding to its provisioned namespace id", () => {
    const result = buildKvNamespacesOverride([{ binding: "CACHE" }], { CACHE: "kv-id-1" });
    expect(result).toEqual([{ binding: "CACHE", id: "kv-id-1" }]);
  });
});

describe("buildR2BucketsOverride", () => {
  it("maps each declared binding to its provisioned bucket name", () => {
    const result = buildR2BucketsOverride([{ binding: "ASSETS" }], { ASSETS: "n-api-assets" });
    expect(result).toEqual([{ binding: "ASSETS", bucket_name: "n-api-assets" }]);
  });
});

describe("buildQueuesOverride", () => {
  it("builds a producer entry for every declared binding", () => {
    const result = buildQueuesOverride({}, [{ binding: "JOBS" }], { JOBS: "n-api-jobs" });
    expect(result.producers).toEqual([{ binding: "JOBS", queue: "n-api-jobs" }]);
    expect(result.consumers).toEqual([]);
  });

  it("builds a consumer entry only for bindings marked consumer: true", () => {
    const result = buildQueuesOverride(
      {},
      [{ binding: "JOBS", consumer: true }, { binding: "EVENTS" }],
      { JOBS: "n-api-jobs", EVENTS: "n-api-events" },
    );
    expect(result.consumers).toEqual([{ queue: "n-api-jobs" }]);
  });

  it("carries forward consumer settings when the base config has exactly one consumer", () => {
    const base = {
      queues: { consumers: [{ queue: "prod-jobs", max_batch_size: 25, max_retries: 3 }] },
    };
    const result = buildQueuesOverride(base, [{ binding: "JOBS", consumer: true }], {
      JOBS: "n-api-jobs",
    });
    expect(result.consumers).toEqual([{ queue: "n-api-jobs", max_batch_size: 25, max_retries: 3 }]);
  });

  it("falls back to no extra settings when the base config has more than one consumer", () => {
    // Which base consumer belongs to which binding is ambiguous without a
    // name/binding key at the consumer level — defaulting rather than
    // guessing wrong is the point of this test.
    const base = {
      queues: {
        consumers: [{ queue: "prod-a", max_batch_size: 1 }, { queue: "prod-b", max_batch_size: 2 }],
      },
    };
    const result = buildQueuesOverride(base, [{ binding: "JOBS", consumer: true }], {
      JOBS: "n-api-jobs",
    });
    expect(result.consumers).toEqual([{ queue: "n-api-jobs" }]);
  });
});

File: internal/deployconfig/deployconfig_test.go
package deployconfig

import (
	"strings"
	"testing"

	"github.com/evatt-labs/kraai/internal/wrangler"
)

func collectWarnings(warnings *[]string) func(string, ...any) {
	return func(format string, args ...any) {
		_ = args
		*warnings = append(*warnings, format)
	}
}

// TestEveryStatefulBindingIsRefused is the finding this package exists for: a
// security audit found ephemeral previews deployed with live production
// bindings for every stateful resource except Hyperdrive. Each key is checked
// individually so adding one to the list without wiring it up fails here.
func TestEveryStatefulBindingIsRefused(t *testing.T) {
	for _, key := range statefulBindingKeys {
		t.Run(key, func(t *testing.T) {
			base := wrangler.Config{"main": "src/index.ts", key: []any{map[string]any{"binding": "X"}}}

			_, err := Build(base, Options{Name: "env-a-api"})
			if err == nil {
				t.Fatalf("%s was inherited without an override or opt-in — the ephemeral Worker "+
					"would have live access to the production resource", key)
			}
			if !strings.Contains(err.Error(), key) {
				t.Fatalf("the error should name %s: %v", key, err)
			}
		})
	}
}

// TestUnknownKeysAreDroppedNotInherited is what makes the allowlist safe
// against Cloudflare shipping binding types faster than kraai learns them.
func TestUnknownKeysAreDroppedNotInherited(t *testing.T) {
	base := wrangler.Config{
		"main": "src/index.ts",
		// A plausible future binding this package has never heard of.
		"hyperdrive_v2_clusters": []any{map[string]any{"binding": "PROD", "id": "prod-cluster"}},
	}

	deployed, err := Build(base, Options{Name: "env-a-api"})
	if err != nil {
		t.Fatalf("Build: %v", err)
	}
	if _, present := deployed["hyperdrive_v2_clusters"]; present {
		t.Fatal("an unrecognised binding was inherited — a new stateful type would leak production by default")
	}
}

// Safe by default is not the same as silent: a dropped key produces a Worker
// that fails at runtime with an unbound variable and nothing explaining why.
func TestUnknownKeysAreReported(t *testing.T) {
	var warnings []string
	base := wrangler.Config{"main": "src/index.ts", "some_new_binding": []any{}}

	if _, err := Build(base, Options{Name: "n", Warn: collectWarnings(&warnings)}); err != nil {
		t.Fatalf("Build: %v", err)
	}
	if len(warnings) == 0 {
		t.Fatal("an unrecognised key was dropped with no warning")
	}
	if !strings.Contains(strings.Join(warnings, "\n"), "unrecognised") {
		t.Fatalf("warnings = %v", warnings)
	}
}

// TestRoutesAreNeverInherited: reassigning a route during an ephemeral deploy
// can redirect real production traffic to a Worker that is about to be
// deleted — a sharper failure than reading production data.
func TestRoutesAreNeverInherited(t *testing.T) {
	for _, key := range neverInheritedKeys {
		t.Run(key, func(t *testing.T) {
			base := wrangler.Config{"main": "src/index.ts", key: "example.com/*"}

			// Even with the opt-in, which is the point.
			deployed, err := Build(base, Options{Name: "env-a-api", UnsafeInheritBindings: true})
			if err != nil {
				t.Fatalf("Build: %v", err)
			}
			if _, present := deployed[key]; present {
				t.Fatalf("%s survived even with unsafeInheritBindings", key)
			}
		})
	}
}

// TestOverridesCannotReintroduceRoutes hardens the guarantee against its own
// callers. Nothing supplies routes through overrides today, but that is a
// property of the callers rather than of this function, and this is the one
// rule here that protects production traffic rather than production data.
func TestOverridesCannotReintroduceRoutes(t *testing.T) {
	var warnings []string
	deployed, err := Build(wrangler.Config{"main": "src/index.ts"}, Options{
		Name:      "env-a-api",
		Overrides: map[string]any{"routes": []any{"prod.example.com/*"}},
		Warn:      collectWarnings(&warnings),
	})
	if err != nil {
		t.Fatalf("Build: %v", err)
	}
	if _, present := deployed["routes"]; present {
		t.Fatal("an override reintroduced routes into the deployed config")
	}
	if len(warnings) == 0 {
		t.Fatal("dropping a route supplied by an override was silent")
	}
}

// A provisioned substitute needs no opt-in: there is nothing
// production-pointing left to inherit once it is replaced.
func TestAnOverrideRemovesTheNeedToOptIn(t *testing.T) {
	base := wrangler.Config{
		"main":         "src/index.ts",
		"d1_databases": []any{map[string]any{"binding": "DB", "database_id": "PROD-ID"}},
	}
	fresh := D1Databases([]Binding{{Binding: "DB"}}, map[string]D1Resource{
		"DB": {Name: "env-a-api-db", ID: "fresh-id"},
	})

	deployed, err := Build(base, Options{Name: "env-a-api", Overrides: map[string]any{"d1_databases": fresh}})
	if err != nil {
		t.Fatalf("a provisioned substitute should not need an opt-in: %v", err)
	}

	entries, _ := deployed["d1_databases"].([]any)
	entry, _ := entries[0].(map[string]any)
	if entry["database_id"] != "fresh-id" {
		t.Fatalf("the deployed config points at %v, not the freshly provisioned database", entry["database_id"])
	}
}

func TestOptInInheritsProductionBindings(t *testing.T) {
	base := wrangler.Config{
		"main":          "src/index.ts",
		"kv_namespaces": []any{map[string]any{"binding": "CACHE", "id": "prod-ns"}},
	}

	deployed, err := Build(base, Options{Name: "env-a-api", UnsafeInheritBindings: true})
	if err != nil {
		t.Fatalf("Build: %v", err)
	}
	if _, present := deployed["kv_namespaces"]; !present {
		t.Fatal("the deliberate opt-in did not inherit the binding")
	}
}

func TestStructuralKeysCarryForward(t *testing.T) {
	base := wrangler.Config{
		"$schema":             "node_modules/wrangler/config-schema.json",
		"main":                "src/index.ts",
		"compatibility_date":  "2026-01-01",
		"compatibility_flags": []any{"nodejs_compat"},
		"observability":       map[string]any{"enabled": true},
		"durable_objects":     map[string]any{"bindings": []any{}},
		"migrations":          []any{map[string]any{"tag": "v1"}},
		"name":                "the-committed-name",
	}

	deployed, err := Build(base, Options{Name: "env-a-api"})
	if err != nil {
		t.Fatalf("Build: %v", err)
	}
	for _, key := range []string{"$schema", "main", "compatibility_date", "compatibility_flags", "observability", "durable_objects", "migrations"} {
		if _, present := deployed[key]; !present {
			t.Errorf("structural key %s was dropped", key)
		}
	}
	// The generated name must win over the committed one, or the ephemeral
	// deploy overwrites the production Worker.
	if deployed["name"] != "env-a-api" {
		t.Fatalf("name = %v, want the generated environment name", deployed["name"])
	}
}

// Durable Objects live inside the deployed script, so a fresh Worker name
// gets fresh storage — there is nothing shared with production to leak, which
// is why they are structural rather than stateful.
func TestDurableObjectsAreNotTreatedAsStateful(t *testing.T) {
	base := wrangler.Config{
		"main":            "src/index.ts",
		"durable_objects": map[string]any{"bindings": []any{map[string]any{"name": "COUNTER", "class_name": "Counter"}}},
	}
	if _, err := Build(base, Options{Name: "env-a-api"}); err != nil {
		t.Fatalf("durable_objects should not require an opt-in: %v", err)
	}
}

func TestVarsAlwaysPresent(t *testing.T) {
	deployed, err := Build(wrangler.Config{"main": "x"}, Options{Name: "n"})
	if err != nil {
		t.Fatal(err)
	}
	vars, ok := deployed["vars"].(map[string]any)
	if !ok {
		t.Fatalf("vars = %v, want an object even when empty", deployed["vars"])
	}
	if len(vars) != 0 {
		t.Fatalf("vars = %v", vars)
	}
}

func TestHyperdriveOnlyWhenProvided(t *testing.T) {
	without, err := Build(wrangler.Config{"main": "x"}, Options{Name: "n"})
	if err != nil {
		t.Fatal(err)
	}
	if _, present := without["hyperdrive"]; present {
		t.Fatal("hyperdrive appeared with none provided")
	}

	with, err := Build(wrangler.Config{"main": "x"}, Options{
		Name:       "n",
		Hyperdrive: []any{map[string]any{"binding": "HYPERDRIVE", "id": "hd-1"}},
	})
	if err != nil {
		t.Fatal(err)
	}
	if _, present := with["hyperdrive"]; !present {
		t.Fatal("hyperdrive was not deployed")
	}
}

func TestQueuesOverride(t *testing.T) {
	base := wrangler.Config{"queues": map[string]any{
		"consumers": []any{map[string]any{"queue": "prod-q", "max_batch_size": float64(10), "max_retries": float64(3)}},
	}}
	bindings := []Binding{{Binding: "JOBS", Consumer: true}, {Binding: "EVENTS"}}
	names := map[string]string{"JOBS": "env-a-jobs", "EVENTS": "env-a-events"}

	got := Queues(base, bindings, names)

	producers, _ := got["producers"].([]any)
	if len(producers) != 2 {
		t.Fatalf("producers = %v, want one per binding", producers)
	}
	consumers, _ := got["consumers"].([]any)
	if len(consumers) != 1 {
		t.Fatalf("consumers = %v, want only the binding the service consumes", consumers)
	}
	consumer, _ := consumers[0].(map[string]any)
	if consumer["queue"] != "env-a-jobs" {
		t.Fatalf("consumer points at %v, not the ephemeral queue", consumer["queue"])
	}
	if consumer["max_retries"] != float64(3) {
		t.Fatalf("the single base consumer's settings were not copied: %v", consumer)
	}
}

// With more than one base consumer, which belongs to which binding is
// genuinely ambiguous — the entries carry no binding key — so settings must
// not be guessed and silently applied to the wrong queue.
func TestQueuesDoesNotGuessBetweenMultipleConsumers(t *testing.T) {
	base := wrangler.Config{"queues": map[string]any{
		"consumers": []any{
			map[string]any{"queue": "a", "max_retries": float64(1)},
			map[string]any{"queue": "b", "max_retries": float64(99)},
		},
	}}

	got := Queues(base, []Binding{{Binding: "JOBS", Consumer: true}}, map[string]string{"JOBS": "env-a-jobs"})

	consumers, _ := got["consumers"].([]any)
	consumer, _ := consumers[0].(map[string]any)
	if _, present := consumer["max_retries"]; present {
		t.Fatalf("settings were guessed from one of two ambiguous consumers: %v", consumer)
	}
	if consumer["queue"] != "env-a-jobs" {
		t.Fatalf("consumer = %v", consumer)
	}
}

// The override builders must preserve declaration order, since bindings are
// positional in the deployed arrays.
func TestOverrideBuildersPreserveOrder(t *testing.T) {
	bindings := []Binding{{Binding: "A"}, {Binding: "B"}, {Binding: "C"}}
	ids := map[string]string{"A": "1", "B": "2", "C": "3"}

	out := KVNamespaces(bindings, ids)
	for i, want := range []string{"A", "B", "C"} {
		entry, _ := out[i].(map[string]any)
		if entry["binding"] != want {
			t.Fatalf("position %d = %v, want %s", i, entry["binding"], want)
		}
	}
}

// TestHyperdriveIsDroppedLoudlyWhenNotReplaced: the committed binding points
// at production, so dropping it is right — but doing it silently leaves a
// Worker that fails on an unbound binding with nothing explaining why.
func TestHyperdriveIsDroppedLoudlyWhenNotReplaced(t *testing.T) {
	var warnings []string
	base := wrangler.Config{
		"main":       "src/index.ts",
		"hyperdrive": []any{map[string]any{"binding": "HYPERDRIVE", "id": "prod-hyperdrive"}},
	}

	deployed, err := Build(base, Options{Name: "env-a-api", Warn: collectWarnings(&warnings)})
	if err != nil {
		t.Fatalf("Build: %v", err)
	}
	if _, present := deployed["hyperdrive"]; present {
		t.Fatal("the production Hyperdrive binding was deployed to an ephemeral Worker")
	}
	if !strings.Contains(strings.Join(warnings, "\n"), "hyperdrive") {
		t.Fatalf("dropping it was silent: %v", warnings)
	}
}

// TestNewerBindingTypesAreRefused pins the keys added against Cloudflare's
// current reference rather than inherited from the JavaScript's stale list.
func TestNewerBindingTypesAreRefused(t *testing.T) {
	for _, key := range []string{"ai_search", "ai_search_namespaces", "tail_consumers", "containers"} {
		t.Run(key, func(t *testing.T) {
			base := wrangler.Config{"main": "src/index.ts", key: []any{map[string]any{"binding": "X"}}}
			if _, err := Build(base, Options{Name: "n"}); err == nil {
				t.Fatalf("%s was not treated as a stateful binding", key)
			}
		})
	}
}
